src/lorawan_mac.erl

```diff
@@ -125,17 +125,19 @@ check_link(DevAddr, FCnt) ->
         [L] ->
             case fcnt_gap(L#link.fcntup, FCnt) of
                 N when N < ?MAX_FCNT_GAP ->
-                    mnesia:dirty_write(links, L#link{fcntup=FCnt}),
+                    NewFCnt = (L#link.fcntup + N) band 16#FFFFFFFF,
+                    mnesia:dirty_write(links, L#link{fcntup = NewFCnt}),
                     {ok, L};
                 BigN ->
                     {error, {fcnt_gap_too_large, BigN}}
             end
     end.
 
 fcnt_gap(A, B) ->
+    C = A band 16#FFFF,
     if
-        A =< B -> B - A;
-        A > B -> 16#FFFF - A + B
+        C > B -> 16#FFFF - C + B;
+        true  -> B - C
```

Comment:
**B** is **FCnt** that comes from **uplink PHYPayload**, it is 16bits. And,we should change **A** to 16bits to get the right Gap(or FCntDiff)